Mathlib/SetTheory/Ordinal/Nimber.lean

```diff
@@ -0,0 +1,362 @@
+/-
+Copyright (c) 2024 Violeta Hernández Palacios. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Violeta Hernández Palacios
+-/
+import Mathlib.SetTheory.Ordinal.Arithmetic
+
+/-!
+# Nimbers
+
+The goal of this file is to define the field of nimbers, constructed as ordinals endowed with new
+arithmetical operations. The nim sum `a + b` is recursively defined as the least ordinal not equal
+to any `a' + b` or `a + b'` for `a' < a` and `b' < b`. The nim product `a * b` is likewise
+recursively defined as the least ordinal not equal to any `a' * b + a * b' + a' * b'` for `a' < a`
+and `b' < b`.
+
+Nim addition arises within the context of impartial games. By the Sprague-Grundy theorem, each
+impartial game is equivalent to some game of nim. If `x ≈ nim o₁` and `y ≈ nim o₂`, then
+`x + y ≈ nim (o₁ + o₂)`, where the ordinals are summed together as nimbers. Unfortunately, the
+nim product admits no such characterization.
+
+## Implementation notes
+
+The nimbers inherit the order from the ordinals - this makes working with minimum excluded values
+much more convenient. However, the fact that nimbers are of characteristic 2 prevents the order from
+interacting with the arithmetic in any nice way.
+
+To reduce API duplication, we opt not to implement operations on `Nimber` on `Ordinal`. The order
+isomorphisms `Ordinal.toNimber` and `Nimber.toOrdinal` allow us to cast between them whenever
+needed.
+
+## Todo
+
+- Define nim multiplication and prove nimbers are a commutative ring.
+- Define nim division and prove nimbers are a field.
+- Show the nimbers are algebraically closed.
```

Comment:
It would also be nice to relate the bitwise `xor` definition on naturals to the addition you've defined here. Something like, doing `xor` on naturals then embedding into nimbers is the same as embedding into nimbers then adding

Reply:
This shouldn't be too difficult to prove using induction plus [`Nat.lt_xor_cases`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Nat/Bitwise.html#Nat.lt_xor_cases), but it does require porting some of the API of `Ordinal.natCast` over, so I'll do it in a separate PR.

Something that might also be good to do in said PR is introduce the notation `∗x` for `Ordinal.toNimber x`.